File: plugins/plugin-windows/src/main/kotlin/com/homecontrol/plugins/windows/protocol/WireMessage.kt
package com.homecontrol.plugins.windows.protocol

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

/**
 * Mirrors `HomeControl.Companion`'s `WireMessage` exactly — same field
 * names, same [SerialName] values for [WireMessageType] (the C# side uses
 * `JsonStringEnumConverter` with default member-name serialization, so the
 * strings here must match the C# enum member names verbatim).
 */
@Serializable
enum class WireMessageType {
    @SerialName("PairRequest") PAIR_REQUEST,
    @SerialName("PairChallenge") PAIR_CHALLENGE,
    @SerialName("PairResult") PAIR_RESULT,
    @SerialName("AuthRequest") AUTH_REQUEST,
    @SerialName("AuthChallenge") AUTH_CHALLENGE,
    @SerialName("AuthResult") AUTH_RESULT,
    @SerialName("Command") COMMAND,
    @SerialName("CommandResult") COMMAND_RESULT,
    @SerialName("Error") ERROR,
}

@Serializable
data class WireMessage(
    val type: WireMessageType,
    val clientPublicKey: String? = null,
    val deviceName: String? = null,
    val serverPublicKey: String? = null,
    val success: Boolean? = null,
    val nonce: String? = null,
    val proofResponse: String? = null,
    val action: String? = null,
    val params: JsonElement? = null,
    val data: JsonElement? = null,
    val message: String? = null,
)

File: windows-companion/src/HomeControl.Companion/Protocol/WireMessage.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace HomeControl.Companion.Protocol;

[JsonConverter(typeof(JsonStringEnumConverter))]
internal enum WireMessageType
{
    PairRequest,
    PairChallenge,
    PairResult,
    AuthRequest,
    AuthChallenge,
    AuthResult,
    Command,
    CommandResult,
    Error,
}

/// <summary>
/// One JSON envelope for every message in both directions — most fields are
/// only meaningful for a subset of <see cref="Type"/> values, kept as one
/// flat shape for a simple wire format rather than a tagged-union hierarchy.
/// </summary>
internal sealed class WireMessage
{
    /// <summary>
    /// The Kotlin side (kotlinx.serialization) uses camelCase property names
    /// by default; plain <c>System.Text.Json</c> defaults to the C# property
    /// names as declared (PascalCase). Every (de)serialization of a
    /// <see cref="WireMessage"/> must use this so the two sides actually
    /// agree on field names — without it, every field silently comes back
    /// null instead of throwing, which is much harder to notice.
    /// </summary>
    public static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
    };

    public WireMessageType Type { get; set; }
    public string? ClientPublicKey { get; set; }
    public string? DeviceName { get; set; }
    public string? ServerPublicKey { get; set; }
    public bool? Success { get; set; }
    public string? Nonce { get; set; }
    public string? ProofResponse { get; set; }
    public string? Action { get; set; }
    public JsonElement? Params { get; set; }
    public JsonElement? Data { get; set; }
    public string? Message { get; set; }

    public static WireMessage PairChallenge(string serverPublicKey) => new()
    {
        Type = WireMessageType.PairChallenge,
        ServerPublicKey = serverPublicKey,
    };

    public static WireMessage PairResult(bool success) => new()
    {
        Type = WireMessageType.PairResult,
        Success = success,
    };

    public static WireMessage AuthChallenge(string nonce) => new()
    {
        Type = WireMessageType.AuthChallenge,
        Nonce = nonce,
    };

    public static WireMessage AuthResult(bool success) => new()
    {
        Type = WireMessageType.AuthResult,
        Success = success,
    };

    public static WireMessage CommandResult(bool success, JsonElement? data) => new()
    {
        Type = WireMessageType.CommandResult,
        Success = success,
        Data = data,
    };

    public static WireMessage Error(string message) => new()
    {
        Type = WireMessageType.Error,
        Message = message,
    };
}
